File: 07-程式語言/code/08-高階語言/_bak/05a-mylisp0/lisp.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <functional>

// 定義一些基本的運算符
std::unordered_map<std::string, std::function<double(double, double)>> ENV = {
    {"+", [](double a, double b) { return a + b; }},
    {"-", [](double a, double b) { return a - b; }},
    {"*", [](double a, double b) { return a * b; }},
    {"/", [](double a, double b) { return a / b; }}
};

// Scheme 解釋器
std::string evaluate(const std::vector<std::string> &exp, const std::unordered_map<std::string, double> &env = {}) {
    if (exp.size() > 0) {
        // 函數調用
        auto op = ENV.at(exp[0]);
        std::vector<double> args;
        for (std::size_t i = 1; i < exp.size(); ++i) {
            args.push_back(evaluate(exp[i], env));
        }
        return op(args[0], args[1]);
    } else if (env.find(exp[0]) != env.end()) {
        // 變數引用
        return env.at(exp[0]);
    } else {
        // 常數
        return exp[0];
    }
}

int main() {
    // 定義一個簡單的 Scheme 表達式
    std::vector<std::string> code = {"+", "2", {"*", "3", "4"}};
    // 解釋和計算結果
    double result = evaluate(code, {});
    std::cout << "Result: " << result << std::endl;

    return 0;
}

File: 05-程式語言/code/09-lisp語言/03-pylisp3/lisp.py
from parse import parse_lisp
import operator as op

isa = isinstance
Symbol = str

class Env(dict):
    def __init__(self, vars, outer=None):
        self.outer = outer
        self.update(vars)

    def find(self, var): # 找到最內層出現的環境變數
        if var in self: return self.get(var)
        elif self.outer is None: raise Exception(var)
        else: return self.outer.find(var)

class Function(object): # 函數定義
    def __init__(self, params, body, env):
        self.params, self.body, self.env = params, body, env
    def __call__(self, *args): 
        if len(args) != len(self.params):
            raise Exception(f'({self.params}) 和 {args} 參數數量不符!')
        fenv = Env(zip(self.params, args), self.env)
        return evaluate(self.body, fenv)

# 定義一些基本的運算符
ENV = Env({ '+': op.add, '-': op.sub, '*': op.mul, '/': op.truediv })

# LISP (Scheme) 解釋器
def evaluate(exp, env={}):
    if isa(exp, list):
        if exp[0] == 'define':
            _, var, value = exp
            env[var] = evaluate(value, env)
        elif exp[0] == 'lambda':
            _, params, body = exp
            return Function(params, body, env)
        else: 
            # 函數調用
            op = evaluate(exp[0], env)
            args = [evaluate(arg, env) for arg in exp[1:]]
            return op(*args)
    elif isa(exp, str) and (env.find(exp)):
        # 變數引用
        return env.find(exp)
    else:
        # 常數
        return exp

def run(blocks):
    env = Env({ '+': op.add, '-': op.sub, '*': op.mul, '/': op.truediv })
    for block in blocks:
        code = parse_lisp(block)
        print(code)
        result = evaluate(code, env)
        print(result)

# 測試 Scheme 解釋器
if __name__ == "__main__":
    # 定義一個 lambda 函數並調用
    run(['((lambda (x y) (+ x y)) 3 4)'])
    
    # 定義一個有名稱的函數並調用
    run([
    "(define add (lambda (x y) (+ x y)))",
    "(add 3 4)"
    ])
